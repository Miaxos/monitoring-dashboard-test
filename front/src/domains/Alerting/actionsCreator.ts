import createActionCreator, {
  ActionsUnion,
} from '../../packages/typedActions/TypedActions';

export const cpuHighLoadStarted = createActionCreator(
  'CPU_HIGH_LOAD_STARTED',
  (date: Date) => ({
    date,
  }),
);

export const cpuHighLoadRecovered = createActionCreator(
  'CPU_HIGH_LOAD_RECOVERED',
  (date: Date) => ({
    date,
  }),
);

const actionsCreators = {
  cpuHighLoadStarted,
  cpuHighLoadRecovered,
};

export type Actions = ActionsUnion<typeof actionsCreators>;
