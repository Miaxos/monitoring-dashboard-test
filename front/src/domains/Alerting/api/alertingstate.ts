import * as ArrayFP from 'fp-ts/es6/Array';
import * as Option from 'fp-ts/es6/Option';
import { createOpaqueAPI } from '@iadvize-oss/opaque-type';
import { pipe } from 'fp-ts/es6/pipeable';
import * as CPUEvent from './alerting/cpu';

type $AlertingState = {
  alertingEvents: CPUEvent.CPUEvent[];
  startedTime: Date;
};

const AlertingState = createOpaqueAPI<'ALERTING_STATE', $AlertingState>(
  'ALERTING_STATE',
);
export type AlertingState = ReturnType<typeof AlertingState.toOpaque>;

/**
 * Create alerting state
 */
export function create(
  alertingEvents: CPUEvent.CPUEvent[] = [],
): AlertingState {
  return AlertingState.toOpaque({
    alertingEvents,
    startedTime: new Date(Date.now()),
  });
}

/**
 * Add an event to the alerting state
 */
export function add(event: CPUEvent.CPUEvent) {
  return (state: AlertingState) =>
    pipe(
      state,
      AlertingState.fromOpaque,
      elt => ({ ...elt, alertingEvents: [event, ...elt.alertingEvents] }),
      AlertingState.toOpaque,
    );
}

/**
 * Get the events from a AlertingState
 * Events aren't sorted by date, but the last in is the first.
 */
export function events(state: AlertingState): CPUEvent.CPUEvent[] {
  return pipe(state, AlertingState.fromOpaque, elt => elt.alertingEvents);
}

/**
 * Get the started time of the alerting
 */
export function startedSince(state: AlertingState): Date {
  return pipe(state, AlertingState.fromOpaque, elt => elt.startedTime);
}

/**
 * It represent the Alerting actual state, either Normal or with an Alert
 */
export enum AlertingStatus {
  NORMAL = 'NORMAL',
  ALERT = 'ALERT',
}

type StatusResponse = {
  status: AlertingStatus;
  date: Option.Option<Date>;
};

/**
 * Get the actual status of the Alerting State and the time when it started.
 */
export function status(state: AlertingState): StatusResponse {
  return pipe(
    state,
    events,
    ArrayFP.head,
    Option.fold(
      () => ({
        status: AlertingStatus.NORMAL,
        date: Option.none,
      }),
      event =>
        pipe(
          event,
          CPUEvent.fold<StatusResponse>({
            onAlert: alertEvent => ({
              status: AlertingStatus.ALERT,
              date: Option.some(CPUEvent.date(alertEvent)),
            }),
            onRecover: recoverEvent => ({
              status: AlertingStatus.NORMAL,
              date: Option.some(CPUEvent.date(recoverEvent)),
            }),
          }),
        ),
    ),
  );
}
