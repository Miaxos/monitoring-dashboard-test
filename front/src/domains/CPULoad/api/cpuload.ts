import * as ArrayFP from 'fp-ts/es6/Array';
import * as Either from 'fp-ts/es6/Either';
import { createOpaqueAPI } from '@iadvize-oss/opaque-type';
import { pipe } from 'fp-ts/es6/pipeable';
import { sequenceS } from 'fp-ts/es6/Apply';

const sequenceSEither = sequenceS(Either.either);

type $CPULoad = Either.Either<
  Error,
  {
    load: number;
    date: Date;
  }
>;

/**
 * A CPU Load can be either Valid or Invalid.
 */
const CPULoad = createOpaqueAPI<'CPU_LOAD', $CPULoad>('CPU_LOAD');

export type CPULoad = ReturnType<typeof CPULoad.toOpaque>;

/**
 * Function to create a CPULoad
 */
export function create(payload: { load: number; date: Date }) {
  return pipe(
    payload,
    elt =>
      sequenceSEither({
        date:
          elt.date.toString() === 'Invalid Date'
            ? Either.left(new Error('Invalid Date'))
            : Either.right(elt.date),
        load: Either.right(elt.load),
      }),
    CPULoad.toOpaque,
  );
}

/**
 * Function to get the data from a CPULoad
 */
export function get(cpuLoad: CPULoad) {
  return pipe(cpuLoad, CPULoad.fromOpaque);
}

/**
 * Get average from an array of CPULoad
 */
export function average(cpuLoads: CPULoad[]) {
  return pipe(
    cpuLoads,
    ArrayFP.map(CPULoad.fromOpaque),
    ArrayFP.rights,
    elt => [elt.length, elt] as const,
    ([length, elt]) =>
      pipe(
        elt,
        ArrayFP.reduce(0, (acc, curr) => curr.load + acc),
        sum => sum / length,
      ),
  );
}
