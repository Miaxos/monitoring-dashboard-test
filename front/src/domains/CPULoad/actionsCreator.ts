import createActionCreator, {
  ActionsUnion,
} from '../../packages/typedActions/TypedActions';
import * as CPULoad from './api/cpuload';

export const cpuLoadRequested = createActionCreator('CPU_LOAD_REQUESTED');

export const cpuLoadFetched = createActionCreator(
  'CPU_LOAD_UPDATED',
  (cpuLoad: CPULoad.CPULoad) => ({
    cpuLoad,
  }),
);

const actionsCreators = {
  cpuLoadFetched,
  cpuLoadRequested,
};

export type Actions = ActionsUnion<typeof actionsCreators>;
