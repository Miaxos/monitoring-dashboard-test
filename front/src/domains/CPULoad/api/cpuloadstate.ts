import { createOpaqueAPI } from '@iadvize-oss/opaque-type';
import * as Option from 'fp-ts/es6/Option';
import { pipe } from 'fp-ts/es6/pipeable';
import * as ArrayFP from 'fp-ts/es6/Array';
import * as CPULoad from './cpuload';

type $CPULoadState = {
  /** Store the cpu loads */
  cpuLoads: CPULoad.CPULoad[];
  /** The loading state is a projection of every task ongoing that are fetching data */
  loadingState: number;
};

const CPULoadState = createOpaqueAPI<'CPU_LOAD_STATE', $CPULoadState>(
  'CPU_LOAD_STATE',
);

export type CPULoadState = ReturnType<typeof CPULoadState.toOpaque>;

/**
 * Function to create a CPULoadState
 */
export function create(cpuLoads: CPULoad.CPULoad[] = []): CPULoadState {
  return CPULoadState.toOpaque({
    cpuLoads,
    loadingState: 0,
  });
}

/**
 * Function to add a CPULoad
 * Actually it's not sorted
 * TODO: Add it sorted.
 */
export function add(cpuLoad: CPULoad.CPULoad) {
  return (state: CPULoadState): CPULoadState =>
    pipe(
      state,
      CPULoadState.fromOpaque,
      elt => ({
        ...elt,
        cpuLoads: [...elt.cpuLoads, cpuLoad],
      }),
      CPULoadState.toOpaque,
    );
}

/**
 * To manage loading projection, we use this fonction to increment
 * our projection.
 */
export function beginLoadCPU(state: CPULoadState): CPULoadState {
  return pipe(
    state,
    CPULoadState.fromOpaque,
    elt => ({ ...elt, loadingState: elt.loadingState + 1 }),
    CPULoadState.toOpaque,
  );
}

/**
 * To manage loading projection, we use this fonction todecrease
 * our projection.
 */
export function endLoadCPU(state: CPULoadState): CPULoadState {
  return pipe(
    state,
    CPULoadState.fromOpaque,
    elt => ({ ...elt, loadingState: elt.loadingState - 1 }),
    CPULoadState.toOpaque,
  );
}

/**
 * Get the last X minutes elements and remove Error Elements
 */
export function getLast(minutes: number) {
  return (state: CPULoadState) =>
    pipe(
      state,
      CPULoadState.fromOpaque,
      elt => ({
        ...elt,
        cpuLoads: pipe(
          elt.cpuLoads,
          ArrayFP.map(CPULoad.get),
          ArrayFP.rights,
          ArrayFP.filterMap(cpu => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            if (Date.now() - cpu.date < minutes * 60 * 1000) {
              return Option.some(CPULoad.create(cpu));
            }
            return Option.none;
          }),
        ),
      }),
      CPULoadState.toOpaque,
    );
}

/**
 * Get cpuLoads
 */
export function cpusLoad(state: CPULoadState) {
  return pipe(state, CPULoadState.fromOpaque, elt => elt.cpuLoads);
}

/**
 * Get average load from a CPULoadState
 */
export function averageLoad(state: CPULoadState): number {
  return pipe(state, cpusLoad, CPULoad.average);
}
