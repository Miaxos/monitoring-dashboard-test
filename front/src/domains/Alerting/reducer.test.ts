import * as AlertingState from './api/alertingstate';
import { cpuHighLoadStarted, cpuHighLoadRecovered } from './actionsCreator';
import AlertingReducer from './reducer';

describe('domains/Alerting/reducer', () => {
  it('should add an alert event', () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementationOnce(() =>
        new Date('2019-05-14T11:01:58.135Z').valueOf(),
      );

    const state = AlertingState.create();
    const action = cpuHighLoadStarted(new Date('2020-03-01T22:24:32.500Z'));
    const reducer = AlertingReducer();

    expect(reducer(state, action)).toMatchSnapshot();
  });

  it('should add an recover event', () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementationOnce(() =>
        new Date('2019-05-14T11:01:58.135Z').valueOf(),
      );

    const state = AlertingState.create();
    const action = cpuHighLoadRecovered(new Date('2020-03-01T22:24:32.500Z'));
    const reducer = AlertingReducer();

    expect(reducer(state, action)).toMatchSnapshot();
  });
});
