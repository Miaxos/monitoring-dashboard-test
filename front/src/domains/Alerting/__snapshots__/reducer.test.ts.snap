// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domains/Alerting/reducer should add an alert event 1`] = `
Object {
  "__OPAQUE_KEY__": "ALERTING_STATE",
  "__OPAQUE__": "__OPAQUE__",
  "value": Object {
    "alertingEvents": Array [
      Object {
        "__OPAQUE_KEY__": "CPU_ALERT",
        "__OPAQUE__": "__OPAQUE__",
        "value": Object {
          "date": 2020-03-01T22:24:32.500Z,
          "type": "CPU_ALERT",
        },
      },
    ],
    "startedTime": 2019-05-14T11:01:58.135Z,
  },
}
`;

exports[`domains/Alerting/reducer should add an recover event 1`] = `
Object {
  "__OPAQUE_KEY__": "ALERTING_STATE",
  "__OPAQUE__": "__OPAQUE__",
  "value": Object {
    "alertingEvents": Array [
      Object {
        "__OPAQUE_KEY__": "CPU_ALERT",
        "__OPAQUE__": "__OPAQUE__",
        "value": Object {
          "date": 2020-03-01T22:24:32.500Z,
          "type": "CPU_RECOVER",
        },
      },
    ],
    "startedTime": 2019-05-14T11:01:58.135Z,
  },
}
`;
